source/lac/matrix_scaling.cc

```diff
@@ -0,0 +1,440 @@
+//
+// ------------------------------------------------------------------------
+
+#include <deal.II/base/memory_consumption.h>
+#include <deal.II/base/numbers.h>
+
+#include <deal.II/lac/block_sparse_matrix.h>
+#include <deal.II/lac/block_sparse_matrix_ez.h>
+#include <deal.II/lac/block_vector.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/matrix_scaling.h>
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/sparse_matrix_ez.h>
+#include <deal.II/lac/vector.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+#include "lac/matrix_scaling.inst"
+
+// Check for unsupported matrix types
+template <typename Matrix>
+constexpr bool
+is_supported_matrix()
+{
+  return std::is_same_v<Matrix, dealii::SparseMatrix<double>> ||
+         std::is_same_v<Matrix, dealii::SparseMatrix<float>> ||
+         std::is_same_v<Matrix, dealii::FullMatrix<double>> ||
+         std::is_same_v<Matrix, dealii::FullMatrix<float>> ||
+         std::is_same_v<Matrix, dealii::BlockSparseMatrix<double>> ||
+         std::is_same_v<Matrix, dealii::BlockSparseMatrix<float>> ||
+         std::is_same_v<Matrix, dealii::SparseMatrixEZ<double>> ||
+         std::is_same_v<Matrix, dealii::SparseMatrixEZ<float>>;
+}
+
+
+MatrixScaling::MatrixScaling(const AdditionalData &control)
+  : control(control)
+  , row_scaling()
+  , column_scaling()
+  , converged(false)
+{}
+
+
+
+template <class Matrix>
+void
+MatrixScaling::scale_matrix(Matrix &matrix)
+{
+  // Check for unsupported matrix types
+  if constexpr (is_supported_matrix<Matrix>())
+    {
+      const auto n_rows = matrix.m();
+      const auto n_cols = matrix.n();
+
+      converged = false;
+
+      row_scaling.reinit(n_rows);
+      column_scaling.reinit(n_cols);
+
+      row_scaling    = 1.0;
+      column_scaling = 1.0;
+
+      switch (control.algorithm)
+        {
+          case MatrixScaling::AdditionalData::ScalingAlgorithm::sinkhorn_knopp:
+            {
+              sk_scaling(matrix,
+                         control.sinkhorn_knopp_parameters.max_iterations);
+            }
+            break;
+          case MatrixScaling::AdditionalData::ScalingAlgorithm::
+            l1_linf_symmetry_preserving:
+            {
+              if (control.l1linf_parameters.start_inf_norm_steps > 0 &&
+                  !converged)
+                linfty_scaling(matrix,
+                               control.l1linf_parameters.start_inf_norm_steps);
+              if (control.l1linf_parameters.l1_norm_steps > 0 && !converged)
+                l1_scaling(matrix, control.l1linf_parameters.l1_norm_steps);
+              if (control.l1linf_parameters.end_inf_norm_steps > 0 &&
+                  !converged)
+                linfty_scaling(matrix,
+                               control.l1linf_parameters.end_inf_norm_steps);
+            }
+            break;
+        }
+    }
+  else
+    {
+      (void)matrix;
+      Assert(false, ExcNotImplemented());
+    }
```

Comment:
This is in the .cc file, so the function is only instantiated for matrix types *you* choose. We don't need to expect users to instantiate this class, so you might as well break compilation for all other matrix types by putting the following at the top of the function.
```
  static_assert (is_supported_matrix<Matrix>());
```
Then you can get rid of the `if constexpr`.